src/CharterManager.sol

```diff
@@ -0,0 +1,253 @@
+}
+
+contract UrnProxy {
+    address immutable public usr;
+
+    constructor(address vat_, address usr_) public {
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+}
+
+contract CharterManager {
+    mapping (address => uint256) public wards;
+    address public implementation;
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "CharterManager/non-authed");
+        _;
+    }
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CharterManagerImp {
+    // --- Data ---
+    mapping (address => uint256) public wards;
+    bytes32 slot1;
+    mapping (address => address) public proxy; // UrnProxy per user
+    mapping (address => mapping (address => uint256)) public can;
+
+    mapping (bytes32 => uint256)                     public gate; // allow only permissioned vaults
+    mapping (bytes32 => uint256)                     public Nib;  // fee percentage for un-permissioned vaults [wad]
+    mapping (bytes32 => mapping(address => uint256)) public nib;  // fee percentage for permissioned vaults    [wad]
+    mapping (bytes32 => mapping(address => uint256)) public line; // debt ceiling for permissioned vaults      [rad]
+
+    address public immutable vat;
+    address public immutable vow;
+
+    // --- Administration ---
+    event File(bytes32 indexed ilk, bytes32 indexed what, bool data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed ilk, address indexed user, bytes32 indexed what, uint256 data);
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        if (what == "gate") gate[ilk] = data;
+        else if (what == "Nib") Nib[ilk] = data;
+        else revert("CharterManager/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+    function file(bytes32 ilk, address user, bytes32 what, uint256 data) external auth {
```

Comment:
nit: `usr` instead of `user` for consistency internally and with dss

Reply:
Done.